### PR TITLE
aws.service-quotas - add new service quotas resource

### PR DESCRIPTION
**Pull request by thisisshi**
This is quite the old PR, I'm definitely OK with removing some of resources in quotas.py for now such as the quota template stuff. Vast majority of the code additions are recorded responses since the service quotas api is quite verbose

Edit: yikes... forgot this was a 10k line diff from all the test recordings...

### CODE REVIEW

**Comment by kapilt** [comment]
this service in particular is likely going to need some manual surgery on recorded responses to keep it size sane.

File: c7n/resources/quotas.py [hunk 1]
```diff
@@ -0,0 +1,477 @@
...
+                    client.exceptions.InvalidResourceStateException,
+                    client.exceptions.ResourceAlreadyExistsException,) as e:
+                error = e
+                continue
+        if error:
+            raise PolicyExecutionError from error
+
+
+@ServiceQuota.action_registry.register('add-to-template')
+class AddToTemplate(Action):
```
Comment (thisisshi) [inline c7n/resources/quotas.py:303]: kind on the fence for this action, docs here: https://docs.aws.amazon.com/servicequotas/latest/userguide/organization-templates.html

**Comment by thisisshi** [comment]
Got this down to ~1k total test recordings, see 88db4d979a7c542cc57102706d5c7de94caaab81 for a commit with the recordings in tact (without the empty responses) (~8k line diff)

**Comment by thisisshi** [comment]
Ugh.. seeing inconsistency on macOS when it comes to reading in placebo recordings, sometimes it passes, sometimes it fails... unsure why

File: c7n/resources/quotas.py (hunk 1 above)
Comment (thisisshi) [reply c7n/resources/quotas.py:306]: @kapilt thoughts?

**Comment by alfredgamulo** [comment]
in general, lgtm. It does feel like `process_resources(self, resources, op)` should be in its own `UpdateRequestTemplateBase`/`ModifyRequestTemplateBase`

File: c7n/resources/quotas.py [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+
+        def get_quotas(client, s):
+            quotas = {}
+            token = None
+            kwargs = {
+                'ServiceCode': s['ServiceCode'],
+                'MaxResults': self.batch_size
+            }
+            while True:
+                if token:
```
Comment (kapilt) [inline c7n/resources/quotas.py:64]: is the lack of bultin pagination the only reason this doesn't work as a detail_spec ? ie. is the set logic/uniqueness here needed?

File: c7n/resources/quotas.py (hunk 1 above)
Comment (kapilt) [reply c7n/resources/quotas.py:306]: i'm also on the fence, its only viable from the organization master account afaics which is tricky to ensure outside of policye execution. if its a one time activity, its generally something we would defer on, but quota mgmt template isn't per se one time, nor is it clear that it applies to existing accounts. tbh it feels more like its own resource re templates.

**Comment by thisisshi** [comment]
@kapilt updated note on organization management account for the add-to-template, remove-from-template, in-template actions/filter

File: tests/test_policy.py [hunk 3]
```diff
@@ -363,6 +363,7 @@ def test_resource_type_empty_metadata(self):
                 continue
             for rk, rv in v.resource_type.__dict__.items():
                 if rk[0].isalnum() and rv is None:
+                    breakpoint()
```
Comment (kapilt) [inline tests/test_policy.py:366]: stray breakpoint

**Comment by thisisshi** [comment]
@jantman I know you were interested in the service quotas stuff, here's a look if you're interested

**Review by kapilt — APPROVED** [review]
thanks, lgtm

**Pull request merged**